### PR TITLE
[bitnami/airflow] Custom Airflow Workers

### PR DESCRIPTION
**Pull request by cccs-cat001**
**Description of the change**
Firstly, I added the ability to specify a secret in kubernetes which holds SSH keys for the dag repo. That way the repo can be private and you won't have to worry about logging in. 
Secondly, I made it possible to clone more than just the dag repo into the workers. 
Thirdly, it's now possible to spawn a custom worker, using whatever image you want, cloning any repos you want, setting any environment variables you want, and mounting any volumes you want.

**Benefits**
This will be useful so that now your workers can be more broad. One issue I was facing with the original worker was that it was very locked down. So this will allow you to make your own worker, give it whatever libraries you want, whatever configurations you want, etc. This change is in addition to having the plain airflow worker too, so now you have multiple workers, each specialized for their own purpose. 

**Possible drawbacks**
None that I'm aware of. 

**Applicable issues**
We use these changes because we host our dags in a private repo, so we needed the ssh keys to be mounted into all the airflow pods it creates. We also use papermill on airflow to schedule parameterized jupyter notebooks,and the simplest solution to that we thought up was to have our notebooks in a git repo (private also) and clone them into the workers similar to how the dags are done. 
Then we came up with this idea to create a custom worker, pulling from a separate queue. These separate workers can do their specific tasks on their queues. But.. you're probably wondering how these workers are pulling from a queue. Well, I'm creating a pull request next in the bitnami-docker-airflow-worker repo with that fix too. [will put the link here](https://github.com/bitnami/bitnami-docker-airflow-worker/pull/2)

**Additional information**

**Checklist** 
- [X] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [X] Variables are documented in the README.md
- [X] Title of the PR starts with chart name (e.g. `[bitnami/chart]`)
- [X] If the chart contains a `values-production.yaml` apart from `values.yaml`, ensure that you implement the changes in both files

:warning: Keep in mind that if you want to make changes to the kubeapps chart, please implement them in the [kubeapps repository](https://github.com/kubeapps/kubeapps/tree/master/chart/kubeapps). This is only a synchronized mirror.

### CODE REVIEW

File: bitnami/airflow/templates/deployment-scheduler.yaml [hunk 1]
```diff
@@ -58,6 +58,8 @@ spec:
           volumeMounts:
             - name: git-cloned-dag-files
               mountPath: /dags
+            - name: dag-git-keys
+              mountPath: /root/.ssh
```
Comment (alemorcuq) [inline bitnami/airflow/templates/deployment-scheduler.yaml]: This volume should only be mounted if `airflow.cloneDagFilesFromGit.sshKeySecret` is set.

File: bitnami/airflow/templates/deployment-scheduler.yaml [hunk 2]
```diff
@@ -201,6 +205,10 @@ spec:
       {{- if .Values.airflow.cloneDagFilesFromGit.enabled }}
         - name: git-cloned-dag-files
           emptyDir: {}
+        - name: dag-git-keys
+          secret:
+            secretName: {{ .Values.airflow.cloneDagFilesFromGit.sshKeySecret }}
+            defaultMode: 0600
```
Comment (alemorcuq) [inline bitnami/airflow/templates/deployment-scheduler.yaml]: This volume should only exist if `airflow.cloneDagFilesFromGit.sshKeySecret` is set.

File: bitnami/airflow/templates/deployment-scheduler.yaml [hunk 3]
```diff
@@ -73,6 +75,8 @@ spec:
           volumeMounts:
             - name: git-cloned-dag-files
               mountPath: /dags
+            - name: dag-git-keys
+              mountPath: /root/.ssh
```
Comment (alemorcuq) [inline bitnami/airflow/templates/deployment-scheduler.yaml]: Same here.

File: bitnami/airflow/templates/deployment-web.yaml [hunk 4]
```diff
@@ -58,6 +58,8 @@ spec:
           volumeMounts:
             - name: git-cloned-dag-files
               mountPath: /dags
+            - name: dag-git-keys
+              mountPath: /root/.ssh
```
Comment (alemorcuq) [inline bitnami/airflow/templates/deployment-web.yaml]: Same here.

File: bitnami/airflow/templates/deployment-web.yaml [hunk 5]
```diff
@@ -73,6 +75,8 @@ spec:
...
```
Comment (alemorcuq) [inline bitnami/airflow/templates/deployment-web.yaml]: Same here.

File: bitnami/airflow/templates/deployment-web.yaml [hunk 6]
```diff
@@ -231,6 +245,10 @@ spec:
         {{- if .Values.airflow.cloneDagFilesFromGit.enabled }}
         - name: git-cloned-dag-files
           emptyDir: {}
+        - name: dag-git-keys
+          secret:
+            secretName: {{ .Values.airflow.cloneDagFilesFromGit.sshKeySecret }}
+            defaultMode: 0600
```
Comment (alemorcuq) [inline bitnami/airflow/templates/deployment-web.yaml]: Same here.

File: bitnami/airflow/templates/deployment-web.yaml [hunk 7]
```diff
@@ -170,6 +174,16 @@ spec:
               {{- else }}
               value: "no"
               {{- end }}
+              {{- if .Values.airflow.pool.enabled }}
+            - name: AIRFLOW_POOL
+              value: "true"
+            - name: AIRFLOW_POOL_NAME
+              value: {{ .Values.airflow.pool.name }}
+            - name: AIRFLOW_POOL_SIZE
+              value: "32"
```
Comment (alemorcuq) [inline bitnami/airflow/templates/deployment-web.yaml]: This shouldn't be hardcoded.

File: bitnami/airflow/templates/deployment-web.yaml [hunk 8]
```diff
@@ -170,6 +174,16 @@ spec:
...
+              {{- if .Values.airflow.pool.enabled }}
+            - name: AIRFLOW_POOL
+              value: "true"
+            - name: AIRFLOW_POOL_NAME
+              value: {{ .Values.airflow.pool.name }}
+            - name: AIRFLOW_POOL_SIZE
+              value: "32" 
+            - name: AIRFLOW_POOL_DESCRIPTION
+              value: {{ .Values.airflow.pool.description }}
+              {{- end }}
```
Comment (alemorcuq) [inline bitnami/airflow/templates/deployment-web.yaml]: The Airflow container should know about this variables. Are you doing a PR there too? I didn't see anything in your message.

File: bitnami/airflow/templates/statefulset-worker.yaml [hunk 9]
```diff
@@ -59,8 +60,29 @@ spec:
           volumeMounts:
             - name: git-cloned-dag-files
               mountPath: /dags
+            - name: dag-git-keys
+              mountPath: /root/.ssh
```
Comment (alemorcuq) [inline bitnami/airflow/templates/statefulset-worker.yaml]: Same comment.

File: bitnami/airflow/templates/statefulset-worker.yaml [hunk 10]
```diff
@@ -200,23 +247,40 @@ spec:
...
+          configMap:
+            name: {{ .Values.airflow.dagsConfigMap }}
+        {{- end }}
+        {{- if .Values.airflow.cloneDagFilesFromGit.enabled }}
+        - name: git-cloned-dag-files
+          emptyDir: {}
+        - name: dag-git-keys
+          secret:
+            secretName: {{ .Values.airflow.cloneDagFilesFromGit.sshKeySecret }}
+            defaultMode: 0600
```
Comment (alemorcuq) [inline bitnami/airflow/templates/statefulset-worker.yaml]: Same comment

File: bitnami/airflow/templates/statefulset-worker.yaml [hunk 11]
```diff
@@ -200,23 +247,40 @@ spec:
...
+          configMap:
+            name: {{ .Values.airflow.configurationConfigMap }}
+        {{- end }}
+        {{- range .Values.additionalRepos.repos }}
+        - name: git-cloned-{{ .name }}
+          emptyDir: {}
+        - name: {{ .name }}-git-keys  
+          secret:
+            secretName: {{ .sshKeySecret }}
+            defaultMode: 0600
```
Comment (alemorcuq) [inline bitnami/airflow/templates/statefulset-worker.yaml]: Same comment.

File: bitnami/airflow/templates/statefulset-worker.yaml [hunk 12]
```diff
@@ -156,6 +201,8 @@ spec:
                   key: airflow-fernetKey
             - name: AIRFLOW_WEBSERVER_HOST
               value: {{ template "airflow.fullname" . }}
+            - name: AIRFLOW_QUEUE
+              value: "default"
```
Comment (alemorcuq) [inline bitnami/airflow/templates/statefulset-worker.yaml]: This value shouldn't be hardcoded.

File: bitnami/airflow/templates/statefulset-worker.yaml [hunk 13]
```diff
@@ -74,6 +96,29 @@ spec:
...
+            - |
+              while true; do
+                  cd /{{ .name }} && git pull origin {{ .branch }}
+                  sleep {{ default "60" .interval }}
+              done
+          volumeMounts:
+            - name: git-cloned-{{ .name }}
+              mountPath: /{{ .name }}
+            - name: {{ .name }}-git-keys
+              mountPath: /root/.ssh
```
Comment (alemorcuq) [inline bitnami/airflow/templates/statefulset-worker.yaml]: Same comment.

File: bitnami/airflow/templates/statefulset-worker.yaml [hunk 14]
```diff
@@ -74,6 +96,29 @@ spec:
           volumeMounts:
             - name: git-cloned-dag-files
               mountPath: /dags
+            - name: dag-git-keys
+              mountPath: /root/.ssh
```
Comment (alemorcuq) [inline bitnami/airflow/templates/statefulset-worker.yaml]: Same comment.

File: bitnami/airflow/templates/statefulset-worker.yaml [hunk 15]
```diff
@@ -59,8 +60,29 @@ spec:
...
+          command:
+            - /bin/bash
+            - -ec
+            - |
+              git clone {{ .repository }} --branch {{ .branch }} /{{ .name }}
+          volumeMounts:
+            - name: git-cloned-{{ .name }}
+              mountPath: /{{ .name }}
+            - name: {{ .name }}-git-keys
+              mountPath: /root/.ssh
```
Comment (alemorcuq) [inline bitnami/airflow/templates/statefulset-worker.yaml]: Same comment.

File: bitnami/airflow/templates/statefulset-worker.yaml [hunk 16]
```diff
@@ -59,8 +60,29 @@ spec:
           volumeMounts:
             - name: git-cloned-dag-files
               mountPath: /dags
+            - name: dag-git-keys
+              mountPath: /root/.ssh
+        {{- end }}
+        {{- if .Values.additionalRepos.enabled }}
+        {{- range .Values.additionalRepos.repos }}
```
Comment (alemorcuq) [inline bitnami/airflow/templates/statefulset-worker.yaml]: This is very inefficient. This is creating a new container for each repo we want to clone. We could clone all the repos using one single container instead.

File: bitnami/airflow/templates/statefulset-worker.yaml [hunk 17]
```diff
@@ -74,6 +96,29 @@ spec:
...
```
Comment (alemorcuq) [inline bitnami/airflow/templates/statefulset-worker.yaml]: Same comment about multiple containers.

**Comment by alemorcuq** [comment]
Hi, @cccs-cat001.

Thank you very much for this, this is an incredible contribution. Since it involves several repositories and makes a big addition, I'd like to proceed slowly with it.

@juan131, @tompizmor, @javsalgar could you take a look at this too?

Thank you again!

File: bitnami/airflow/templates/deployment-web.yaml (hunk 8 above)
Comment (cccs-cat001) [reply bitnami/airflow/templates/deployment-web.yaml]: Sorry yes, I've got a pull request for the airflow image too - https://github.com/bitnami/bitnami-docker-airflow/pull/19

**Comment by cccs-cat001** [comment]
> I don't think I understand very well the reasoning behind the additional StatefulSets for different configurations, could you explain it, maybe with some examples?

I made them stateful sets because I don't really know kubernetes that well. All I did was copy the `statefulset-worker.yaml` file, and modified it to work for a list of workers. If there is a more reasonable way to make custom workers without it being a stateful set I'm open to suggestions. 

> I also think that, instead of adding an array for extra repos and another one for extra workers, maybe we should just use an array in the first place.

So you mean define the default worker also in this `workers` list? 

> Changes in the `values.yaml` should be replicated in the `values-production.yaml`. If there are any different default settings, they should be documented in the README.

Sorry thought I did this before, guess I didn't save

> And last, you should increase the chart version.

I'm not sure what to increase to. 4.1.16 or 4.0.17?

File: bitnami/airflow/templates/statefulset-worker.yaml (hunk 16 above)
Comment (Rkbhasan1) [reply bitnami/airflow/templates/statefulset-worker.yaml]: I am not sure.

File: bitnami/airflow/templates/statefulset-worker.yaml (hunk 17 above)
Comment (Rkbhasan1) [reply bitnami/airflow/templates/statefulset-worker.yaml]: I don't know.

File: bitnami/airflow/templates/statefulset-worker.yaml (hunk 16 above)
Comment (cccs-cat001) [reply bitnami/airflow/templates/statefulset-worker.yaml]: I changed it so it spawns only one pod, and then clones all the repos into that pod, and mounts a volume for each ssh key. 

In saying that, I realize that if all the repos require an ssh key, it will not work correctly. You can only use an ssh key for one repo. So we do need all the containers to be created - one for each repo.

File: bitnami/airflow/README.md [hunk 18]
```diff
@@ -149,6 +155,26 @@ The following tables lists the configurable parameters of the Kafka chart and th
 | `metrics.image.pullPolicy`                | Image pull policy                                                                                    | `IfNotPresent`                                               |
 | `metrics.image.pullSecrets`               | Specify docker-registry secret names as an array                                                     | `[]` (does not add image pull secrets to deployed pods)      |
 | `metrics.podAnnotations`                  | Additional annotations for Metrics exporter                                                          | `{prometheus.io/scrape: "true", prometheus.io/port: "9112"}` |
+| `additionalRepos.enabled`                 | Enables the use of additional repos in the default Airflow Workers                                   | `false` 							  |
+| `additionalRepos.repos[o].name`	    | Descriptive name of a repo to sync			 					   | `nil`							  |
```
Comment (juan131) [inline bitnami/airflow/README.md]: Typo: `additionalRepos.repos[o].name` should be `additionalRepos.repos[0].name`
You used `o` instead of `0`

File: bitnami/airflow/README.md [hunk 19]
```diff
@@ -149,6 +155,26 @@ The following tables lists the configurable parameters of the Kafka chart and th
...
+| `workers[0].image.registry`    	    | Registry for this workers image									   | `nil`							  |
+| `workers[0].image.repository`		    | Name of this worker image										   | `nil`							  | 
+| `workers[0].image.tag`		    | Tag for this worker image										   | `nil`							  | 
+| `workers[0].image.pullPolicy` 	    | Pull policy for this worker image									   | `nil`							  | 
+| `workers[0].additionalRepos.enabled`	    | Enables the use of additional repos for this worker						   | `false`							  |
+| `workers[0].additionalRepos.repos[0].name` | Descriptive name of a repo to sync with this worker 	 					   | `nil`							  |
+| `workers[0].additionalRepos.repos[0].repository` | Link to the additional repository to sync with this worker					   | `nil` 							  |
+| `workers[0].additionalRepos.repos[0].branch` | Branch from repository to sync with this worker							   | `nil`							  | 
+| `workers[0].additionalRepos.repos[0].interval` | Interval to pull the repository for this worker							   | 60 							  |
+| `workers[0].additionalRepos.repos[0].sshKeySecret` | Kubernetes secret containing SSH keys for this repo						   | `nil`							  |
```
Comment (juan131) [inline bitnami/airflow/README.md]: I would name this parameter `workers[0].additionalRepos.repos[0].sshKeysSecretName`

File: bitnami/airflow/values.yaml [hunk 20]
```diff
@@ -218,7 +227,7 @@ ingress:
   ## If certManager is set to true, annotation kubernetes.io/tls-acme: "true" will automatically be set
   annotations:
   #  kubernetes.io/ingress.class: nginx
-
+
```
Comment (juan131) [inline bitnami/airflow/values.yaml]: Remove this extra space

File: bitnami/airflow/values.yaml [hunk 21]
```diff
@@ -365,9 +374,38 @@ metrics:
...
+    #  tag:
+    #  pullPolicy:
+    # additionalRepos:
+      # enabled: false
+      # repos:
+      # - name:
+      #   repository:
+      #   branch:
+      #   interval:
+      #   sshKeySecret:
```
Comment (juan131) [inline bitnami/airflow/values.yaml]: What about the following approach:
```yaml
      ## SSH keys to clone the repo
      ## 
      # sshKeys:
      ## Name of existing secret containing the SSH keys to clone the repo
      ## When provided, `workers[0].additionalRepos.repos[0].sshKeys` will be ignored
      ##
      # sshKeysSecretName:
```
Then you can create a **ssh-keys-secrets.yaml** manifest like the one below:
```yaml
{{- range .Values.workers }}
{{- if .additionalRepos.enabled }}
{{- range .additionalRepos.repos }}
{{- if empty .sshKeysSecretName }}
apiVersion: v1
kind: Secret
metadata:
  name: {{ .name }}
  labels:
    app.kubernetes.io/name: {{ include "airflow.name" $ }}
    helm.sh/chart: {{ include "airflow.chart" $ }}
    app.kubernetes.io/instance: {{ $.Release.Name }}
    app.kubernetes.io/managed-by: {{ $.Release.Service }}
data: 
  ssh-keys: {{ .sshKeys | b64enc }}
---
{{- end }}
{{- end }}
{{- end }}
{{- end }}
```

File: bitnami/airflow/templates/statefulset-worker-list.yaml [hunk 22]
```diff
@@ -0,0 +1,306 @@
+{{- range $.Values.workers }}
```
Comment (juan131) [inline bitnami/airflow/templates/statefulset-worker-list.yaml]: I don't think the `$` is necessary here

File: bitnami/airflow/values.yaml (hunk 21 above)
Comment (cccs-cat001) [reply bitnami/airflow/values.yaml]: I'm not sure I understand here. Are you saying we could include the sshkeys directly here, and then have it generate the secrets with those keys, as opposed to how I have it set up now where it's assumed that the ssh keys are already a secret, and you tell it which secret they're in?

File: bitnami/airflow/templates/statefulset-worker-list.yaml (hunk 22 above)
Comment (cccs-cat001) [reply bitnami/airflow/templates/statefulset-worker-list.yaml]: I would agree

File: bitnami/airflow/values.yaml (hunk 21 above)
Comment (juan131) [reply bitnami/airflow/values.yaml]: Exactly. You support both taken the ssh keys from an already existing secret, and passing them in the values so you create the secret

File: bitnami/airflow/README.md [hunk 23]
```diff
@@ -149,6 +155,21 @@ The following tables lists the configurable parameters of the Kafka chart and th
 | `metrics.image.pullPolicy`                | Image pull policy                                                                                    | `IfNotPresent`                                               |
 | `metrics.image.pullSecrets`               | Specify docker-registry secret names as an array                                                     | `[]` (does not add image pull secrets to deployed pods)      |
 | `metrics.podAnnotations`                  | Additional annotations for Metrics exporter                                                          | `{prometheus.io/scrape: "true", prometheus.io/port: "9112"}` |
+| `additionalRepos.enabled`                 | Enables the use of additional repos in the default Airflow Workers                                   | `false` 							  |
```
Comment (juan131) [inline bitnami/airflow/README.md]: I think this parameter is not necessary anymore

File: bitnami/airflow/README.md [hunk 24]
```diff
@@ -233,8 +254,42 @@ airflow.cloneDagFilesFromGit.enabled=true
...
+You can specify additional repositories to sync with your worker for additional files for your DAGs to use, or whatever pleases you. You can specify multiple. To do so: 
+
+```console
+additionalRepos.enabled=true
+additionalRepos.repos[0].name=additional-repo
+additionalRepos.repos[0].repository=https://github.com/USERNAME/REPOSITORY
+additionalRepos.repos[0].branch=master
+additioanlRepos.repos[0].interval=60
+additioanlRepos.repos[0].sshKeySecretName=KUBERNETES_SECRET
+```
```
Comment (juan131) [inline bitnami/airflow/README.md]: This isn't true anymore now we're using a single list for workers, is it?

File: bitnami/airflow/README.md (hunk 23 above)
Comment (juan131) [reply bitnami/airflow/README.md]: The `workerImage` parameter is also deprecated now

File: bitnami/airflow/values.yaml [hunk 25]
```diff
@@ -56,30 +56,6 @@ schedulerImage:
...
-  ## ref: https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/
-  ##
-  # pullSecrets:
-  #   - myRegistryKeySecretName
-
-  ## Set to true if you would like to see extra information on logs
-  ## It turns BASH and NAMI debugging in minideb
-  ## ref:  https://github.com/bitnami/minideb-extras/#turn-on-bash-debugging
-  debug: false
-
```
Comment (juan131) [inline bitnami/airflow/values.yaml]: Update the **values-production.yaml**

File: bitnami/airflow/values.yaml (hunk 21 above)
Comment (cccs-cat001) [reply bitnami/airflow/values.yaml]: I think this assumes that you're definitely supplying ssh keys to every repo eh? I made an enabled option to turn it on in case you don't need ssh keys.

File: bitnami/airflow/templates/deployment-scheduler.yaml [hunk 26]
```diff
@@ -58,6 +58,10 @@ spec:
           volumeMounts:
             - name: git-cloned-dag-files
               mountPath: /dags
+            {{- if .Values.airflow.cloneDagFilesFromGit.sshKeySecret }}
```
Comment (juan131) [inline bitnami/airflow/templates/deployment-scheduler.yaml]: `sshKeySecret` -> `sshKeySecretName`

Comment (juan131) [reply bitnami/airflow/templates/deployment-scheduler.yaml]: You need to apply these change several times on this manifest

File: bitnami/airflow/templates/deployment-web.yaml [hunk 27]
```diff
@@ -58,6 +58,10 @@ spec:
           volumeMounts:
             - name: git-cloned-dag-files
               mountPath: /dags
+            {{- if .Values.airflow.cloneDagFilesFromGit.sshKeySecret }}
```
Comment (juan131) [inline bitnami/airflow/templates/deployment-web.yaml]: `sshKeySecret` -> `sshKeySecretName`

Comment (juan131) [reply bitnami/airflow/templates/deployment-web.yaml]: You need to apply these change several times on this manifest

File: bitnami/airflow/values-production.yaml [hunk 28]
```diff
@@ -155,7 +129,7 @@ airflow:
     ## Option to force users to specify a password. That is required for 'helm upgrade' to work properly.
     ## If it is not forced, a random password will be generated.
     ##
-    forcePassword: true
+    forcePassword: false
```
Comment (juan131) [inline bitnami/airflow/values-production.yaml]: Revert this changes in the production values

File: bitnami/airflow/values-production.yaml [hunk 29]
```diff
@@ -205,7 +186,7 @@ service:
 ##
 ingress:
   ## Set to true to enable ingress record generation
-  enabled: true
+  enabled: false
```
Comment (juan131) [inline bitnami/airflow/values-production.yaml]: Revert this changes in the production values

**Comment by juan131** [comment]
Hi @cccs-cat001 

It seems that sth definitely went wrong when you updated your branch (as you can see there are changes related to other charts)

If I were you, I'd close this PR and create a new one including your latest changes, since I think it's going be much easier than fixing all the conflicts you have at the moment.
